File: dotnet/VM/Values/VariableValue.cs
#nullable enable

namespace SimpleStackVM
{
    public struct VariableValue : IValue
    {
        #region Fields
        public readonly string Value;

        public bool IsLabel => this.Value.Length > 1 && this.Value[0] == ':';

        public string TypeName => "variable";
        #endregion

        #region Constructor
        public VariableValue(string value)
        {
            this.Value = value;
        }
        #endregion

        #region Methods
        public int CompareTo(IValue? other)
        {
            if (other == null) return 1;
            if (other is VariableValue otherString)
            {
                return this.Value.CompareTo(otherString.Value);
            }

            return 1;
        }

        public override string ToString() => this.Value;
        #endregion
    }
}

File: ts/src/values/variableValue.ts
import { CompareResult, IValue } from "./ivalues";

export class VariableValue implements IValue
{
    public readonly value: string;

    public get isLabel() { return this.value.length > 0 && this.value[0] === ':'; }

    constructor (value: string)
    {
        this.value = value;
    }

    public typename() { return 'variable'; }
    public toString() { return this.value; }

    public compareTo(other: IValue): CompareResult
    {
        if (!(other instanceof VariableValue))
        {
            return 1;
        }

        const diff = this.value.localeCompare(other.value);
        if (diff == 0) { return 0; }
        if (diff < 0) { return -1; }
        return 1;
    }
}